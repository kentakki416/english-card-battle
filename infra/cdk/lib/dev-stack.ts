import * as cdk from 'aws-cdk-lib';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as ecs from 'aws-cdk-lib/aws-ecs';
import * as ecr from 'aws-cdk-lib/aws-ecr';
import * as elbv2 from 'aws-cdk-lib/aws-elasticloadbalancingv2';
import { Construct } from 'constructs';

export class DevStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // VPCの作成
    const vpc = new ec2.Vpc(this, 'VPC', {
      vpcName: 'english-card-battle-dev-vpc',
      maxAzs: 2,
      natGateways: 1,
      subnetConfiguration: [
        {
          name: 'Public',
          subnetType: ec2.SubnetType.PUBLIC,
          cidrMask: 24,
        },
        {
          name: 'Private',
          subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS,
          cidrMask: 24,
        },
      ],
    });

    // ECRリポジトリの作成
    const repository = new ecr.Repository(this, 'Repository', {
      repositoryName: 'english-card-battle-dev',
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      imageScanOnPush: true,
      lifecycleRules: [
        {
          maxImageCount: 5,
          tagStatus: ecr.TagStatus.ANY,
        },
        {
          maxImageAge: cdk.Duration.days(1),
          tagStatus: ecr.TagStatus.UNTAGGED,
        },
      ],
    });

    // ECSクラスターの作成
    const cluster = new ecs.Cluster(this, 'Cluster', {
      clusterName: 'english-card-battle-dev-cluster',
      vpc,
    });

    // セキュリティグループの作成
    const albSg = new ec2.SecurityGroup(this, 'AlbSecurityGroup', {
      vpc,
      securityGroupName: 'english-card-battle-dev-alb-sg',
      description: 'Security group for ALB',
    });

    const ecsSg = new ec2.SecurityGroup(this, 'EcsSecurityGroup', {
      vpc,
      securityGroupName: 'english-card-battle-dev-ecs-sg',
      description: 'Security group for ECS tasks',
    });

    // セキュリティグループルールの設定
    albSg.addIngressRule(
      ec2.Peer.anyIpv4(),
      ec2.Port.tcp(80),
      'Allow HTTP traffic'
    );

    ecsSg.addIngressRule(
      albSg,
      ec2.Port.tcp(80),
      'Allow traffic from ALB'
    );

    // ALBの作成
    const alb = new elbv2.ApplicationLoadBalancer(this, 'ALB', {
      vpc,
      internetFacing: true,
      securityGroup: albSg,
      vpcSubnets: {
        subnetType: ec2.SubnetType.PUBLIC,
      },
    });

    // ターゲットグループの作成
    const targetGroup = new elbv2.ApplicationTargetGroup(this, 'TargetGroup', {
      vpc,
      port: 80,
      protocol: elbv2.ApplicationProtocol.HTTP,
      targetType: elbv2.TargetType.IP,
      healthCheck: {
        path: '/',
        interval: cdk.Duration.seconds(30),
      },
    });

    // リスナーの作成
    alb.addListener('Listener', {
      port: 80,
      defaultTargetGroups: [targetGroup],
    });

    // ECSタスク定義の作成
    const taskDefinition = new ecs.FargateTaskDefinition(this, 'TaskDefinition', {
      memoryLimitMiB: 512,
      cpu: 256,
    });

    // コンテナの定義
    taskDefinition.addContainer('AppContainer', {
      image: ecs.ContainerImage.fromRegistry('nginx:latest'),
      portMappings: [{ containerPort: 80 }],
      logging: ecs.LogDrivers.awsLogs({
        streamPrefix: 'english-card-battle-dev',
        logRetention: cdk.aws_logs.RetentionDays.THREE_DAYS,
      }),
    });

    // ECSサービスの作成
    const service = new ecs.FargateService(this, 'Service', {
      cluster,
      taskDefinition,
      desiredCount: 1,
      securityGroups: [ecsSg],
      assignPublicIp: false,
      vpcSubnets: {
        subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS,
      },
    });

    // ターゲットグループにサービスを登録
    service.attachToApplicationTargetGroup(targetGroup);

    // 出力の設定
    new cdk.CfnOutput(this, 'AlbDnsName', {
      value: alb.loadBalancerDnsName,
      description: 'ALB DNS Name',
    });

    new cdk.CfnOutput(this, 'EcrRepositoryUri', {
      value: repository.repositoryUri,
      description: 'ECR Repository URI',
    });
  }
} 
