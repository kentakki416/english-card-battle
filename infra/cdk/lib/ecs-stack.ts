import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { Vpc, IpAddresses, SubnetType, SecurityGroup, Peer, Port } from 'aws-cdk-lib/aws-ec2';
import { ApplicationLoadBalancer } from 'aws-cdk-lib/aws-elasticloadbalancingv2';
import { Role, ServicePrincipal, ManagedPolicy } from 'aws-cdk-lib/aws-iam';
import {
  FargateTaskDefinition,
  ContainerImage,
  LogDriver,
  Cluster,
  FargateService,
  Protocol,
} from 'aws-cdk-lib/aws-ecs';


export class EcsStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const sampleVpc = new Vpc(this, 'SampleVpc', {
      ipAddresses: IpAddresses.cidr('10.100.0.0/16'),
      maxAzs: 3,
      natGateways: 1,
      subnetConfiguration: [
        {
          cidrMask: 24,
          name: 'Public',
          subnetType: SubnetType.PUBLIC,
        },
        {
          cidrMask: 24,
          name: 'Private',
          subnetType: SubnetType.PRIVATE_WITH_EGRESS,
        },
      ],
    });

    const securityGroupForAlb = new SecurityGroup(this, 'SgAlb', {
      vpc: sampleVpc,
      allowAllOutbound: false,
    });
    securityGroupForAlb.addIngressRule(Peer.anyIpv4(), Port.tcp(80));
    // MEMO: 証明書作成後は以下に変更
    // securityGroupForAlb.addIngressRule(Peer.anyIpv4(), Port.tcp(443));
    securityGroupForAlb.addEgressRule(Peer.anyIpv4(), Port.allTcp());

    const albForApp = new ApplicationLoadBalancer(this, 'Alb', {
      vpc: sampleVpc,
      internetFacing: true,
      securityGroup: securityGroupForAlb,
      vpcSubnets: sampleVpc.selectSubnets({
        subnetGroupName: 'Public',
      }),
    });
    const securityGroupForFargate = new SecurityGroup(this, 'SgFargate',{
        vpc: sampleVpc,
        allowAllOutbound: false,
    });
    securityGroupForFargate.addIngressRule(securityGroupForAlb, Port.tcp(80));
    securityGroupForFargate.addEgressRule(Peer.anyIpv4(),Port.allTcp());

    const executionRole = new Role(this, 'EcsTaskExecutionRole', {
      assumedBy: new ServicePrincipal('ecs-tasks.amazonaws.com'),
      managedPolicies: [
        ManagedPolicy.fromAwsManagedPolicyName(
          'service-role/AmazonECSTaskExecutionRolePolicy'
        ),
      ],
    });
    const serviceTaskRole = new Role(this, 'EcsServiceTaskRole', {
      assumedBy: new ServicePrincipal('ecs-tasks.amazonaws.com'),
      managedPolicies: [
        ManagedPolicy.fromAwsManagedPolicyName('AmazonSSMFullAccess'),
      ],
    });

    const taskDefinition = new FargateTaskDefinition(this, `TaskDefinition`, {
        cpu: 256,
        memoryLimitMiB: 512,
        executionRole: executionRole,
        taskRole: serviceTaskRole,
      }
    );

    taskDefinition.addContainer('api', {
      image: ContainerImage.fromRegistry("amazon/amazon-ecs-sample"),
      logging: LogDriver.awsLogs({
        streamPrefix: `samplePrefixName`,
      })
    }).addPortMappings({
      containerPort: 80,
      hostPort: 80,
      protocol: Protocol.TCP,
    });

    const cluster = new Cluster(this, 'Cluster', {
      vpc: sampleVpc,
      containerInsights: true,
      // containerInsightsV2: ContainerInsights.ENABLED,
    });

    const fargateService = new FargateService(this,'FargateService',{
      cluster,
      vpcSubnets: sampleVpc.selectSubnets({ subnetGroupName: 'Private' }),
      securityGroups: [securityGroupForFargate],
      taskDefinition: taskDefinition,
      desiredCount: 1,
      maxHealthyPercent: 200,
      minHealthyPercent: 50,
      enableExecuteCommand: true,
    });

    const albListener = albForApp.addListener('AlbListener', {
      port: 80,
    });
    const fromAppTargetGroup = albListener.addTargets('FromAppTargetGroup',{
        port: 80,
        targets: [fargateService],
    });
    new cdk.CfnOutput(this, 'LoadBalancerDNS', {
      value: albForApp.loadBalancerDnsName,
    });
  }
}
