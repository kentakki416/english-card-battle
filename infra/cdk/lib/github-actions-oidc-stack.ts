import * as cdk from "aws-cdk-lib"
import { Construct } from "constructs"

import { GitHubActionsOidcConstruct } from "./constructs/github-actions-oidc"

export class GitHubActionsOidcStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: cdk.StackProps) {
    super(scope, id, props)

    new GitHubActionsOidcConstruct(this, "GitHubActionsOidc", {
      githubOwner: "kentakki416",
      githubRepository: "english-card-battle"
    })
  }
}
