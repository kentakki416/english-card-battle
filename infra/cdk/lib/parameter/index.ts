import { devParameters } from './dev'
import { prdParameters } from './prd'
import { stgParameters } from './stg'
import { Environment, EnvironmentParameters } from './types'

export function getEnvironmentParameters(environment: Environment): EnvironmentParameters {
  switch (environment) {
    case 'dev':
      return devParameters
    case 'stg':
      return stgParameters
    case 'prd':
      return prdParameters
    default:
      return devParameters
  }
}

export * from './types'
