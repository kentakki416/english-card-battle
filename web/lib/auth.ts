import GoogleProvider from "next-auth/providers/google"

export const authOptions = {
    providers: [
        GoogleProvider({
            clientId: process.env.GOOGLE_OAUTH_CLIENT_ID || "",
            clientSecret: process.env.GOOGLE_OAUTH_CLIENT_SECRET || "",
        }),
    ],
    secret: process.env.NEXTAUTH_SECRET,
    callbacks: {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        async jwt({ token, account }: any) {
            // Persist the OAuth access_token to the token right after signin
            if (account) {
              token.accessToken = account.access_token
            }
            return token
          },
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          async session({ session, token }: any) {
            // Send properties to the client, like an access_token from a provider.
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (session as any).accessToken = (token as any).accessToken
            return session
          }
    }
} 
