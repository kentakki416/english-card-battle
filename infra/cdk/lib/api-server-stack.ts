import * as cdk from 'aws-cdk-lib'
import { Construct } from 'constructs'

import { EcrConstruct } from './constructs/ecr'
import { EcsConstruct } from './constructs/ecs'
import { VpcConstruct } from './constructs/vpc'
import { getEnvironmentParameters, Environment } from './parameter'

interface ApiServerStackProps extends cdk.StackProps {
  environment: Environment;
}

export class ApiServerStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: ApiServerStackProps) {
    super(scope, id, props)

    // 環境パラメータの取得
    const parameters = getEnvironmentParameters(props.environment)

    // VPC Constructの作成
    const vpcConstruct = new VpcConstruct(this, 'VpcConstruct', {
      vpcCidr: parameters.vpc.vpcCidr,
      maxAzs: parameters.vpc.maxAzs,
      natGateways: parameters.vpc.natGateways,
      subnetConfiguration: parameters.vpc.subnetConfiguration,
    })

    // ECR Constructの作成
    new EcrConstruct(this, 'EcrConstruct', {
      repositoryName: parameters.ecr.repositoryName,
    })

    // ECS Constructの作成
    new EcsConstruct(this, 'EcsConstruct', {
      vpc: vpcConstruct.vpc,
      cpu: parameters.ecs.cpu,
      memoryLimitMiB: parameters.ecs.memoryLimitMiB,
      containerPort: parameters.ecs.containerPort,
      hostPort: parameters.ecs.hostPort,
      desiredCount: parameters.ecs.desiredCount,
      maxHealthyPercent: parameters.ecs.maxHealthyPercent,
      minHealthyPercent: parameters.ecs.minHealthyPercent,
      enableExecuteCommand: parameters.ecs.enableExecuteCommand,
    })
  }
}
