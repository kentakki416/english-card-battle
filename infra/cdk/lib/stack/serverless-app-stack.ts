import * as cdk from 'aws-cdk-lib'
import { Construct } from 'constructs'

import { Api } from '../constructs/api'
import { Auth } from '../constructs/auth'
import { DataStore } from '../constructs/datastore'

export class ServerLessAppStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props)
    const auth = new Auth(this, 'Auth')

    const datastore = new DataStore(this, 'DataStore')

    new Api(this, 'Api', {
      userPool: auth.userPool,
      sampleTable: datastore.sampleTable
    })
  }
}

