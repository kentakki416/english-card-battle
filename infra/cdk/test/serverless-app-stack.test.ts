import * as cdk from "aws-cdk-lib"
import { Template } from "aws-cdk-lib/assertions"

import { ServerLessAppStack } from "../lib/serverless-app-stack"

describe(__filename, () => {

  describe("スナップショットテスト", () => {
    it("スナップショットが一致する", () => {
        const app = new cdk.App()
        const stack = new ServerLessAppStack(app, "TestStack", {})
        const template = Template.fromStack(stack)

        expect(template.toJSON()).toMatchSnapshot()
    })
  })
})
