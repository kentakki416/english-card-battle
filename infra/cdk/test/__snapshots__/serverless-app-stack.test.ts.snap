// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/Users/s16865/workspace/local/product/english-card-battle/infra/cdk/test/serverless-app-stack.test.ts スナップショットテスト スナップショットが一致する 1`] = `
{
  "Outputs": {
    "ApiRestApiEndpoint4DEFB5CC": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "ApiRestApiE35FEAFE",
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region",
            },
            ".",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "ApiRestApiDeploymentStagev157F58EAC",
            },
            "/",
          ],
        ],
      },
    },
    "AuthUserPoolClientId8216BF9A": {
      "Value": {
        "Ref": "AuthUserPoolUserPoolClientBB863FCC",
      },
    },
    "AuthUserPoolIdC0605E59": {
      "Value": {
        "Ref": "AuthUserPool8115E87F",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ApiRestApiANY1340C355": {
      "Properties": {
        "AuthorizationType": "COGNITO_USER_POOLS",
        "AuthorizerId": {
          "Ref": "ApiUserPoolsAuthroizer2967CC9C",
        },
        "HttpMethod": "ANY",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "ApiSampleLamdba9DBA2CBF",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Fn::GetAtt": [
            "ApiRestApiE35FEAFE",
            "RootResourceId",
          ],
        },
        "RestApiId": {
          "Ref": "ApiRestApiE35FEAFE",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "ApiRestApiANYApiPermissionTestStackApiRestApi914E81ACANY48CD34B4": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "ApiSampleLamdba9DBA2CBF",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "ApiRestApiE35FEAFE",
              },
              "/",
              {
                "Ref": "ApiRestApiDeploymentStagev157F58EAC",
              },
              "/*/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ApiRestApiANYApiPermissionTestTestStackApiRestApi914E81ACANYED3364A6": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "ApiSampleLamdba9DBA2CBF",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "ApiRestApiE35FEAFE",
              },
              "/test-invoke-stage/*/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ApiRestApiAccount598A1DA6": {
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "ApiRestApiE35FEAFE",
      ],
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": [
            "ApiRestApiCloudWatchRoleB7879BFC",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "ApiRestApiCloudWatchRoleB7879BFC": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "ApiRestApiDeploymentDC05A82D382718a6dd343826668c6cf2cce6b8e6": {
      "DependsOn": [
        "ApiRestApiproxyANY7D1D2F15",
        "ApiRestApiproxyOPTIONS14872AA4",
        "ApiRestApiproxy95CB4350",
        "ApiRestApiANY1340C355",
        "ApiRestApiOPTIONSB925735A",
      ],
      "Properties": {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": {
          "Ref": "ApiRestApiE35FEAFE",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "ApiRestApiDeploymentStagev157F58EAC": {
      "DependsOn": [
        "ApiRestApiAccount598A1DA6",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "ApiRestApiDeploymentDC05A82D382718a6dd343826668c6cf2cce6b8e6",
        },
        "RestApiId": {
          "Ref": "ApiRestApiE35FEAFE",
        },
        "StageName": "v1",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "ApiRestApiE35FEAFE": {
      "Properties": {
        "Name": "RestApi",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "ApiRestApiOPTIONSB925735A": {
      "Properties": {
        "ApiKeyRequired": false,
        "AuthorizationType": "NONE",
        "HttpMethod": "OPTIONS",
        "Integration": {
          "IntegrationResponses": [
            {
              "ResponseParameters": {
                "method.response.header.Access-Control-Allow-Headers": "'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token,X-Amz-User-Agent'",
                "method.response.header.Access-Control-Allow-Methods": "'OPTIONS,GET,PUT,POST,DELETE,PATCH,HEAD'",
                "method.response.header.Access-Control-Allow-Origin": "'*'",
                "method.response.header.Access-Control-Max-Age": "'300'",
              },
              "StatusCode": "204",
            },
          ],
          "RequestTemplates": {
            "application/json": "{ statusCode: 200 }",
          },
          "Type": "MOCK",
        },
        "MethodResponses": [
          {
            "ResponseParameters": {
              "method.response.header.Access-Control-Allow-Headers": true,
              "method.response.header.Access-Control-Allow-Methods": true,
              "method.response.header.Access-Control-Allow-Origin": true,
              "method.response.header.Access-Control-Max-Age": true,
            },
            "StatusCode": "204",
          },
        ],
        "ResourceId": {
          "Fn::GetAtt": [
            "ApiRestApiE35FEAFE",
            "RootResourceId",
          ],
        },
        "RestApiId": {
          "Ref": "ApiRestApiE35FEAFE",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "ApiRestApiproxy95CB4350": {
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": [
            "ApiRestApiE35FEAFE",
            "RootResourceId",
          ],
        },
        "PathPart": "{proxy+}",
        "RestApiId": {
          "Ref": "ApiRestApiE35FEAFE",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "ApiRestApiproxyANY7D1D2F15": {
      "Properties": {
        "AuthorizationType": "COGNITO_USER_POOLS",
        "AuthorizerId": {
          "Ref": "ApiUserPoolsAuthroizer2967CC9C",
        },
        "HttpMethod": "ANY",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "ApiSampleLamdba9DBA2CBF",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "ApiRestApiproxy95CB4350",
        },
        "RestApiId": {
          "Ref": "ApiRestApiE35FEAFE",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "ApiRestApiproxyANYApiPermissionTestStackApiRestApi914E81ACANYproxyB0CF38AA": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "ApiSampleLamdba9DBA2CBF",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "ApiRestApiE35FEAFE",
              },
              "/",
              {
                "Ref": "ApiRestApiDeploymentStagev157F58EAC",
              },
              "/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ApiRestApiproxyANYApiPermissionTestTestStackApiRestApi914E81ACANYproxy33B067D6": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "ApiSampleLamdba9DBA2CBF",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "ApiRestApiE35FEAFE",
              },
              "/test-invoke-stage/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ApiRestApiproxyOPTIONS14872AA4": {
      "Properties": {
        "ApiKeyRequired": false,
        "AuthorizationType": "NONE",
        "HttpMethod": "OPTIONS",
        "Integration": {
          "IntegrationResponses": [
            {
              "ResponseParameters": {
                "method.response.header.Access-Control-Allow-Headers": "'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token,X-Amz-User-Agent'",
                "method.response.header.Access-Control-Allow-Methods": "'OPTIONS,GET,PUT,POST,DELETE,PATCH,HEAD'",
                "method.response.header.Access-Control-Allow-Origin": "'*'",
                "method.response.header.Access-Control-Max-Age": "'300'",
              },
              "StatusCode": "204",
            },
          ],
          "RequestTemplates": {
            "application/json": "{ statusCode: 200 }",
          },
          "Type": "MOCK",
        },
        "MethodResponses": [
          {
            "ResponseParameters": {
              "method.response.header.Access-Control-Allow-Headers": true,
              "method.response.header.Access-Control-Allow-Methods": true,
              "method.response.header.Access-Control-Allow-Origin": true,
              "method.response.header.Access-Control-Max-Age": true,
            },
            "StatusCode": "204",
          },
        ],
        "ResourceId": {
          "Ref": "ApiRestApiproxy95CB4350",
        },
        "RestApiId": {
          "Ref": "ApiRestApiE35FEAFE",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "ApiSampleLamdba9DBA2CBF": {
      "DependsOn": [
        "ApiSampleLamdbaServiceRoleDefaultPolicyD1AD3FE7",
        "ApiSampleLamdbaServiceRole5047C520",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "a5810395091e8eada051a2fab1947e66a27603d8b72a0b4faf6768bdd2dd7e1a.zip",
        },
        "Environment": {
          "Variables": {
            "TABLE_NAME": {
              "Ref": "DataStoreSampleTable06F53423",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "ApiSampleLamdbaServiceRole5047C520",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "ApiSampleLamdbaServiceRole5047C520": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ApiSampleLamdbaServiceRoleDefaultPolicyD1AD3FE7": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "DataStoreSampleTable06F53423",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ApiSampleLamdbaServiceRoleDefaultPolicyD1AD3FE7",
        "Roles": [
          {
            "Ref": "ApiSampleLamdbaServiceRole5047C520",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ApiUserPoolsAuthroizer2967CC9C": {
      "Properties": {
        "IdentitySource": "method.request.header.Authorization",
        "Name": "TestStackApiUserPoolsAuthroizer259BCC33",
        "ProviderARNs": [
          {
            "Fn::GetAtt": [
              "AuthUserPool8115E87F",
              "Arn",
            ],
          },
        ],
        "RestApiId": {
          "Ref": "ApiRestApiE35FEAFE",
        },
        "Type": "COGNITO_USER_POOLS",
      },
      "Type": "AWS::ApiGateway::Authorizer",
    },
    "AuthUserPool8115E87F": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AccountRecoverySetting": {
          "RecoveryMechanisms": [
            {
              "Name": "verified_phone_number",
              "Priority": 1,
            },
            {
              "Name": "verified_email",
              "Priority": 2,
            },
          ],
        },
        "AdminCreateUserConfig": {
          "AllowAdminCreateUserOnly": false,
        },
        "AutoVerifiedAttributes": [
          "email",
        ],
        "DeletionProtection": "ACTIVE",
        "EmailVerificationMessage": "The verification code to your new account is {####}",
        "EmailVerificationSubject": "Verify your new account",
        "SmsVerificationMessage": "The verification code to your new account is {####}",
        "UsernameAttributes": [
          "email",
        ],
        "VerificationMessageTemplate": {
          "DefaultEmailOption": "CONFIRM_WITH_CODE",
          "EmailMessage": "The verification code to your new account is {####}",
          "EmailSubject": "Verify your new account",
          "SmsMessage": "The verification code to your new account is {####}",
        },
      },
      "Type": "AWS::Cognito::UserPool",
      "UpdateReplacePolicy": "Retain",
    },
    "AuthUserPoolUserPoolClientBB863FCC": {
      "Properties": {
        "AllowedOAuthFlows": [
          "code",
        ],
        "AllowedOAuthFlowsUserPoolClient": true,
        "AllowedOAuthScopes": [
          "email",
          "profile",
          "openid",
        ],
        "CallbackURLs": [
          "https://example.com",
        ],
        "ExplicitAuthFlows": [
          "ALLOW_ADMIN_USER_PASSWORD_AUTH",
          "ALLOW_REFRESH_TOKEN_AUTH",
        ],
        "GenerateSecret": false,
        "LogoutURLs": [
          "https://example.com",
        ],
        "SupportedIdentityProviders": [
          "COGNITO",
        ],
        "UserPoolId": {
          "Ref": "AuthUserPool8115E87F",
        },
      },
      "Type": "AWS::Cognito::UserPoolClient",
    },
    "AuthUserPoolUserPoolDomain4EEAE7CF": {
      "Properties": {
        "Domain": "dev-mastering-cdk-serverless-app",
        "UserPoolId": {
          "Ref": "AuthUserPool8115E87F",
        },
      },
      "Type": "AWS::Cognito::UserPoolDomain",
    },
    "DataStoreSampleTable06F53423": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "Replicas": [
          {
            "Region": {
              "Ref": "AWS::Region",
            },
          },
        ],
      },
      "Type": "AWS::DynamoDB::GlobalTable",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
